src/asl/asl-graph.ts

```diff
@@ -2037,13 +2046,102 @@ export namespace ASLGraph {
    *
    * https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html#amazon-states-language-intrinsic-functions-escapes
    */
-  export function escapeFormatString(literal: ASLGraph.LiteralValue) {
-    if (typeof literal.value === "string") {
-      return literal.value.replace(/[\}\{\'}]/g, "\\$&");
+  export function escapeFormatLiteral(literal: ASLGraph.LiteralValue) {
+    if (ASLGraph.isLiteralString(literal)) {
+      return escapeFormatLiteralString(literal);
     } else {
       return literal.value;
     }
   }
+
+  /**
+   * Turns any literal value in a string or intrinsic functions that can generate a string representation.
+   *
+   * ```ts
+   * literalValue("value") // => "value"
+   * literalValue(1) // => "1"
+   * literalValue({ a: 1 }) // => "{ \"a\": "1" }""
+   * literalValue({ "a.$": "$.var" }) => intrinsicFormat('{ "a": {} }', JsonToString("$.var"))
+   * literalValue(["$.var"]) => intrinsicFormat('[{}]', JsonToString("$.var"))
+   * ```
+   */
+  export function stringifyLiteral(
+    value: ASLGraph.LiteralValue
+  ): ASLGraph.IntrinsicFunction | ASLGraph.LiteralValue<string> {
+    if (ASLGraph.isLiteralNull(value)) {
+      return ASLGraph.literalValue("null");
+    } else if (ASLGraph.isLiteralArray(value)) {
+      return stringifyLiteralArray(value);
+    } else if (ASLGraph.isLiteralObject(value)) {
+      return stringifyLiteralObject(value);
+    }
+    return ASLGraph.isLiteralString(value)
+      ? value
+      : ASLGraph.literalValue(String(value.value));
+
+    function stringifyLiteralObject(
+      value: ASLGraph.LiteralValue<Record<string, ASLGraph.LiteralValueType>>
+    ): ASLGraph.IntrinsicFunction | ASLGraph.LiteralValue<string> {
+      if (!value.containsJsonPath) {
+        return ASLGraph.literalValue(JSON.stringify(value.value), false);
+      }
+      const entries = Object.entries(value.value).map(([key, value]) => {
+        if (key.endsWith(".$")) {
+          // assuming the these values are
+          return [
+            key.substring(0, key.length - 2),
+            ASLGraph.intrinsicJsonToString(ASLGraph.jsonPath(value as string)),
+          ] as const;
+        }
+        return [key, stringifyLiteral(ASLGraph.literalValue(value))] as const;
+      });
+
+      if (entries.every(([_, value]) => ASLGraph.isLiteralValue(value))) {
+        // just stringify the original value.
+        return ASLGraph.literalValue(JSON.stringify(value.value));
+      }
```

Comment:
Can you explain how this works? Looks like a nice optimization but why can you serialize the value.value here

Reply:
This works off the assumption that the stringify functions only return LiteralValue when all of the children are also literals.

If one value is a jsonPath, nested stringify will start to return interinsicFunctions. 

Will document.